docs/typescript.md

```diff
@@ -0,0 +1,132 @@
+
+## Generating TypeScript
+
+Given a model represented in `morphir-ir.json`, you can generate a TypeScript
+representation by running:
+
+    morphir-elm gen --input morphir-ir.json --output ./generated --target=TypeScript
+
+Note that at present only the types are converted. Data values and functions
+are not.
+
+You can generate a TypeScript representation of the Morphir IR itself by
+running this in the morphir-elm repo:
+
+    morphir-elm make ./morphir-make --types-only
+    morphir-elm gen --input=morphir-ir.json --output=./generated --target=TypeScript
+
+## Using the generated types
+
+The TypeScript backend outputs a top-level module per package, which your own
+code should import. The namespaces correspond with the package and module names
+in the IR. Only namespaces and symbols marked as public will be exported in the
+TypeScript API.
+
+For example, you can use the `IR` types from the `Morphir` package like this:
+
+    import { Morphir } from './generated/Morphir'
+   
+    const myName: Morphir.IR.Name.Name = ["this", "is", "a", "great", "name"]
+
+Internally the types are mapped to JavaScript types. You will find this in
+`generated/morphir/ir/Name.ts`:
+
+    export type Name = Array<string>
+
+You benefit from all the usual TypeScript type checking. For example, a Path
+must be a list of Name instances, so this example will not compile:
+
+    import { Morphir } from './generated/Morphir'
+   
+    const myName: Morphir.IR.Path.Path = "This is the wrong type."
+
+You should see an error when compiling this:
+
+    test.ts:3:7 - error TS2322: Type 'string' is not assignable to type 'Path'.
+
+Most Morphir types correspond directly to JavaScript types. The
+[JSON mapping](https://github.com/finos/morphir-elm/blob/main/docs/json-mapping.md)
+gives a useful reference. There are some special cases, which are documented below.
+
+### Type mapping details
+
+#### Dict
+
+A `Morphir.SDK.Dict.Dict K V` maps to a TypeScript `Array<K,V>`. We plan to
+change to using
+[Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map).
+
+#### Custom types
+
+We follow the example
+["Tagged Union Types in TypeScript"](https://mariusschulz.com/blog/tagged-union-types-in-typescript)
+to implement custom types.
+
+Each type variant is a TypeScript `interface`, with a `kind` and maybe some
+fields: `arg1`, `arg2`, `arg3` and so on.
+
+You need to construct these manually. Here's an example using the Morphir IR
+`Literal` type:
+
+    import { Morphir } from './generated/Morphir'
+
+    let myBool: Morphir.IR.Literal.BoolLiteral = {
+        kind: "BoolLiteral",
+        arg1: true,
+    }
+
+    let myString: Morphir.IR.Literal.StringLiteral = {
+        kind: "StringLiteral",
+        arg1: "This is a wonderful string",
+    }
+
+    let myLiteral: Morphir.IR.Literal.Literal = myString;
+
+Note that this example would not actually compile as the `Morphir.IR.Literal`
```

Comment:
I also think it's especially misleading, since we use native types like `string` and `bool` in our generated code, and because the best way to get a string in TypeScript would be
```
let myString: string = "wonderful string"
```

Bringing `Morphir.IR.Literal.BoolLiteral` and `Morphir.IR.Literal.StringLiteral` into the conversation is going to make people think that they need to use those types whenever they want a boolean or a string. And so far as I know, they don't need to.

Things might get more complicated when we start supporting functions and values. But for now, I don't think they would need to use those types in order to interact with our code.

Reply:
Hopefully we can rewrite this section anyway, now that we have constructor functions.

Reply:
Good point that it's a bit confusing. I chose it because it's a simple Custom type that doesn't recursively depend on other types. What would be a better example? I guess Constructor is one option. And the FullName type from the Reference model is another option.